client/my-sites/plugins/plugins-discovery-page/index.jsx

```diff
@@ -24,63 +24,88 @@ import usePlugins from '../use-plugins';
+const UpgradeNudge = ( {
+	selectedSite,
+	sitePlan,
+	isVip,
+	jetpackNonAtomic,
+	siteSlug,
+	paidPlugins,
+} ) => {
+	const hasInstallPurchasedPlugins = useSelector( ( state ) =>
+		siteHasFeature( state, selectedSite?.ID, WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS )
+	);
+	const requiredPlansPurchasedPlugins = useSelector( ( state ) =>
+		getPlansForFeature( state, selectedSite?.ID, WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS )
+	);
+	const requiredPlansAllPlugins = useSelector( ( state ) =>
+		getPlansForFeature( state, selectedSite?.ID, FEATURE_INSTALL_PLUGINS )
+	);
+
 	const translate = useTranslate();
+
 	const eligibleForProPlan = useSelector( ( state ) =>
 		isEligibleForProPlan( state, selectedSite?.ID )
 	);
 
-	if ( ! selectedSite?.ID || ! sitePlan || isVip || jetpackNonAtomic ) {
-		return null;
-	}
-	const isLegacyPlan = isBlogger( sitePlan ) || isPersonal( sitePlan ) || isPremium( sitePlan );
-	const checkoutPlan = eligibleForProPlan && ! isLegacyPlan ? 'pro' : 'business';
-	const bannerURL = `/checkout/${ siteSlug }/${ checkoutPlan }`;
-	const plan = findFirstSimilarPlanKey( sitePlan.product_slug, {
-		type: TYPE_BUSINESS,
-	} );
+	const requiredPlans = useSelector( ( state ) =>
+		getPlansForFeature( state, selectedSite?.ID, WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS )
+	);
 
-	const title =
-		eligibleForProPlan && ! isLegacyPlan
-			? translate( 'Upgrade to the Pro plan to install plugins.' )
-			: translate( 'Upgrade to the Business plan to install plugins.' );
+	// Whether to show an upgrade banner specific to premium plugins.
+	const lowerPlanAvailable =
+		! hasInstallPurchasedPlugins &&
+		requiredPlansPurchasedPlugins[ 0 ] !== requiredPlansAllPlugins[ 0 ];
 
-	return (
-		<UpsellNudge
-			event="calypso_plugins_browser_upgrade_nudge"
-			showIcon={ true }
-			href={ bannerURL }
-			feature={ FEATURE_INSTALL_PLUGINS }
-			plan={ plan }
-			title={ title }
-		/>
-	);
-};
+	let bannerURL;
+	let feature;
+	let plan;
+	let title;
 
-const UpgradeNudgePaid = ( props ) => {
-	const translate = useTranslate();
+	if ( paidPlugins && ! hasInstallPurchasedPlugins && lowerPlanAvailable ) {
+		if ( ! requiredPlans ) {
+			return null;
+		}
 
-	const requiredPlans = useSelector( ( state ) =>
-		getPlansForFeature( state, props.selectedSite?.ID, WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS )
-	);
+		const requiredPlan = getPlan( requiredPlans[ 0 ] );
 
-	if ( ! requiredPlans ) {
+		bannerURL = `/checkout/${ siteSlug }/${ requiredPlan.getPathSlug() }`;
+		feature = WPCOM_FEATURES_INSTALL_PURCHASED_PLUGINS;
+		plan = requiredPlan.getStoreSlug();
+		title = translate( 'Upgrade to the %(planName)s plan to install premium plugins.', {
+			textOnly: true,
+			args: { planName: requiredPlan.getTitle() },
+		} );
+	} else if (
+		( paidPlugins && ! hasInstallPurchasedPlugins && ! lowerPlanAvailable ) ||
+		( ! paidPlugins && ( hasInstallPurchasedPlugins || lowerPlanAvailable ) )
```

Comment:
The logic is hard to follow, but basically, there are potentially 2 distinct upsells:

1. one for unlocking the feature to install "paid plugins" (the top banner)
2. one for the feature to install "free plugins" (the 2nd banner)

- If it's a Jetpack site, we should show NO banners.
- If the banners would be duplicates, we should only show the top banner, because the upsell applies to both free and paid plugins.
- If the banners to upsell free and paid plugins are different, we should show both.
- If we have the ability to install paid plugins, but not free plugins (Starter plan), we should show the banner upselling free plugin installs (the 2nd banner), but not the top banner.
- If you can install free plugins, you can install paid plugins, so that banner scenario doesn't exist.

Reply:
Thanks for laying this out, I think this is the condition I was missing:

- If we have the ability to install paid plugins, but not free plugins (Starter plan), we should show the banner upselling free plugin installs (the 2nd banner), but not the top banner.

Reply:
> If the banners would be duplicates, we should only show the top banner, because the upsell applies to both free and paid plugins.

This was what I wasn't understanding!